src/units.cpp

```diff
@@ -0,0 +1,24 @@
+#include "json.h"
+#include "units.h"
+
+template<>
+void units::volume::serialize( JsonOut &jsout ) const
+{
+    if( value_ % 1000 == 0 ) {
+        jsout.write( string_format( "%d L", value_ ) );
+    } else {
+        jsout.write( string_format( "%d ml", value_ ) );
+    }
+}
+
+template<>
+void units::mass::serialize( JsonOut &jsout ) const
+{
+    if( value_ % 1000000 == 0 ) {
+        jsout.write( string_format( "%d kg", value_ ) );
+    } else if( value_ % 1000 == 0 ) {
+        jsout.write( string_format( "%d g", value_ ) );
+    } else {
+        jsout.write( string_format( "%d mg", value_ ) );
+    }
+}
```

Comment:
See https://stackoverflow.com/questions/25594644/warning-specialization-of-template-in-different-namespace
tl;dr it seems to be a gcc 5.4 bug.
I think you want something like:
```
namespace units
{
template<>
void volume::serialize( JsonOut &jsout ) const
{
    if( value_ % 1000 == 0 ) {
        jsout.write( string_format( "%d L", value_ ) );
    } else {
        jsout.write( string_format( "%d ml", value_ ) );
    }
}

template<>
void mass::serialize( JsonOut &jsout ) const
{
    if( value_ % 1000000 == 0 ) {
        jsout.write( string_format( "%d kg", value_ ) );
    } else if( value_ % 1000 == 0 ) {
        jsout.write( string_format( "%d g", value_ ) );
    } else {
        jsout.write( string_format( "%d mg", value_ ) );
    }
}
};
```